Bug: `tofu login` can panic with "close of closed channel" during OAuth flow

### Community note

> [!TIP]
> 👋 Hi there, OpenTofu community! The OpenTofu team prioritizes issues based on upvotes. Please make sure to upvote this issue and describe how it affects you in detail in the comments to show your support.


### OpenTofu Version

```shell
OpenTofu v1.12.0-dev on darwin_arm64
```

### OpenTofu Configuration Files

```hcl
...tofu config...
```
N/A — this bug is in the CLI OAuth login flow, not in `.tf` configuration files.


### Debug Output

The panic can be reproduced with the following minimal test that mirrors the exact goroutine structure of `interactiveGetTokenByCode`. It is **not** a test of `LoginCommand` itself, but isolates the channel mechanics at lines 454/479/527:

```go
// Place inside internal/command/ (package command) and run:
// go test -race -run TestReproDoubleClose ./internal/command/
func TestReproDoubleClose(t *testing.T) {
    for i := 0; i < 200; i++ {
        func() {
            defer func() {
                if r := recover(); r != nil {
                    t.Errorf("panic (no recover in production): %v", r)
                }
            }()

            codeCh := make(chan string)
            triggerServerError := make(chan struct{})

            // close #1 — mirrors HTTP callback handler (line 454)
            go func() {
                codeCh <- "auth-code"
                close(codeCh)
            }()

            // close #2 — mirrors server.Serve() error path (line 479)
            go func() {
                <-triggerServerError
                close(codeCh) // panics if codeCh already closed
            }()

            go func() {
                time.Sleep(time.Microsecond * time.Duration(i%3))
                close(triggerServerError)
            }()

            // close #3 — mirrors ShutdownCh select branch (line 527)
            shutdownCh := make(chan struct{}, 1)
            select {
            case <-shutdownCh:
                close(codeCh)
            case <-codeCh:
            }
        }()
    }
}
```

Running this produces:

```
=== RUN   TestReproDoubleClose
panic: close of closed channel

goroutine 24 [running]:
...internal/command/login_race_repro_test.go:81 +0x80
created by ...interactiveGetTokenByCode.func1
        internal/command/login.go:479
FAIL    github.com/opentofu/opentofu/internal/command   1.020s
```

### Expected Behavior

`tofu login` should handle concurrent shutdown signals and OAuth callback responses gracefully, without panicking.


### Actual Behavior

`tofu login` panics with:
<img width="839" height="727" alt="Image" src="https://github.com/user-attachments/assets/21c028de-a028-4535-b172-a3d2f82a806c" />

### Steps to Reproduce

**Scenario A — Ctrl+C races with a valid OAuth callback:**

1. Run `tofu login app.terraform.io`
2. Complete the OAuth flow in the browser (click "Approve")
3. Press Ctrl+C at the exact moment the browser redirects back to `localhost`

**Scenario B — server error + Ctrl+C (more reliably reproducible):**

1. Block all TCP ports in the range used by the OAuth callback server (default 10000–10010)
2. Run `tofu login app.terraform.io`
3. Press Ctrl+C while the command is running

---

### Additional Context


- This race was introduced in PR #1074, which added Ctrl+C support to the OAuth login flow by adding the third `close(codeCh)` site (line 527), without accounting for the existing  close sites at lines 454 and 479.
- The bug is a timing race, so it will not occur on every run. It is most likely to surface under slow network conditions or when the OAuth callback and a Ctrl+C signal arrive simultaneously.


### References

- PR #1074 — "shutdown OAuth2 callback server when user hits Ctrl+C" (introduced close #3)

--- a/internal/command/login.go
+++ b/internal/command/login.go
@@ -19,6 +19,7 @@ import (
 	"net/url"
 	"path/filepath"
 	"strings"
+	"sync"
 
 	tfe "github.com/hashicorp/go-tfe"
 	"github.com/mitchellh/cli"
@@ -418,8 +419,14 @@ func (c *LoginCommand) interactiveGetTokenByCode(ctx context.Context, hostname s
 	}
 
 	// codeCh will allow our temporary HTTP server to transmit the OAuth code
-	// to the main execution path that follows.
-	codeCh := make(chan string)
+	// to the main execution path that follows. It is buffered so the handler
+	// can send without blocking even if the main goroutine has already moved on.
+	// Only the main goroutine may close codeCh, after all producers have stopped.
+	codeCh := make(chan string, 1)
+	// serverErrCh carries any unexpected error from server.Serve so that the
+	// main goroutine can handle it without a shared-state race on diags.
+	serverErrCh := make(chan error, 1)
+	var wg sync.WaitGroup
 	server := &http.Server{
 		Handler: http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
 			log.Printf("[TRACE] login: request to callback server")
@@ -442,12 +449,17 @@ func (c *LoginCommand) interactiveGetTokenByCode(ctx context.Context, hostname s
 				return
 			}
 
-			log.Printf("[TRACE] login: request contains an authorization code")
-
-			// Send the code to our blocking wait below, so that the token
-			// fetching process can continue.
-			codeCh <- gotCode
-			close(codeCh)
+			// Non-blocking send: only the first callback request succeeds.
+			// Duplicate or concurrent requests are rejected with 400 so that
+			// the handler never blocks on the already-full buffered channel.
+			select {
+			case codeCh <- gotCode:
+				log.Printf("[TRACE] login: request contains an authorization code")
+			default:
+				log.Printf("[WARN] login: ignoring duplicate callback request")
+				resp.WriteHeader(400)
+				return
+			}
 
 			log.Printf("[TRACE] login: returning response from callback server")
 
@@ -460,21 +472,13 @@ func (c *LoginCommand) interactiveGetTokenByCode(ctx context.Context, hostname s
 		}),
 	}
 	panicHandler := logging.PanicHandlerWithTraceFn()
-	go func() {
+	wg.Go(func() {
 		defer panicHandler()
 		err := server.Serve(listener)
 		if err != nil && err != http.ErrServerClosed {
-			diags = diags.Append(tfdiags.Sourceless(
-				tfdiags.Error,
-				"Can't start temporary login server",
-				fmt.Sprintf(
-					"The login process uses OAuth, which requires starting a temporary HTTP server on localhost. However, no TCP port numbers between %d and %d are available to create such a server.",
-					clientConfig.MinPort, clientConfig.MaxPort,
-				),
-			))
-			close(codeCh)
+			serverErrCh <- err
 		}
-	}()
+	})
 
 	oauthConfig := &oauth2.Config{
 		ClientID:    clientConfig.ID,
@@ -509,7 +513,6 @@ func (c *LoginCommand) interactiveGetTokenByCode(ctx context.Context, hostname s
 	view.WaitingForHostSignal()
 
 	var code string
-	var ok bool
 	select {
 	case <-c.ShutdownCh:
 		diags = diags.Append(
@@ -519,22 +522,36 @@ func (c *LoginCommand) interactiveGetTokenByCode(ctx context.Context, hostname s
 				"Current command was aborted by the calling code.",
 			),
 		)
-		code, ok = "", true
+    if err := server.Shutdown(ctx); err != nil {
+		log.Printf("[WARN] login: callback server shutdown failed: %s", err)
+	}
+		wg.Wait()
 		close(codeCh)
-	case code, ok = <-codeCh:
-	}
-
-	if !ok {
-		// If we got no code at all then the server wasn't able to start
-		// up, so we'll just give up.
 		return nil, diags
+	case serveErr := <-serverErrCh:
+		// The server failed to start up, so we'll just give up.
+		// No need to call Shutdown here: the server never started accepting
+		// requests, so there are no in-flight handler goroutines to wait for.
+		log.Printf("[ERROR] login: callback server error: %s", serveErr)
+		diags = diags.Append(tfdiags.Sourceless(
+			tfdiags.Error,
+			"Can't start temporary login server",
+			fmt.Sprintf(
+				"The login process uses OAuth, which requires starting a temporary HTTP server on localhost. However, no TCP port numbers between %d and %d are available to create such a server.",
+				clientConfig.MinPort, clientConfig.MaxPort,
+			),
+		))
+		wg.Wait()
+		close(codeCh)
+		return nil, diags
+	case code = <-codeCh:
 	}
 
-	if err := server.Close(); err != nil {
-		// The server will close soon enough when our process exits anyway,
-		// so we won't fuss about it for right now.
+	if err := server.Shutdown(ctx); err != nil {
 		log.Printf("[WARN] login: callback server can't shut down: %s", err)
 	}
+	wg.Wait()
+	close(codeCh)
 
 	if code == "" {
 		// empty code is not possible in happy path as it is validated in the HTTP handler of our callback server

--- a/internal/command/login_test.go
+++ b/internal/command/login_test.go
@@ -11,6 +11,7 @@ import (
 	"path/filepath"
 	"strings"
 	"testing"
+	"time"
 
 	"github.com/opentofu/opentofu/internal/command/workdir"
 	"github.com/opentofu/opentofu/internal/terminal"
@@ -368,4 +369,162 @@ func TestLogin(t *testing.T) {
 			t.Fatalf("missing expected error message\nwant: %s\nfull output:\n%s", want, got)
 		}
 	}, false))
+}
+
+// TestLoginOAuthCallbackRace verifies that no panic, deadlock, or data race
+// occurs when ShutdownCh fires concurrently with the OAuth callback completing.
+//
+// Before the fix, interactiveGetTokenByCode had multiple goroutines calling
+// close(codeCh), which could panic with "close of closed channel". The fix
+// uses server.Shutdown() to wait for all in-flight handler goroutines to
+// finish before closing codeCh, ensuring the main goroutine is the sole owner
+// of the channel's lifetime.
+//
+// Run with the race detector to catch any remaining data races:
+//
+//	go test -race -run TestLoginOAuthCallbackRace ./internal/command/
+func TestLoginOAuthCallbackRace(t *testing.T) {
+	s := httptest.NewServer(oauthserver.Handler)
+	defer s.Close()
+
+	const iterations = 200
+
+	for i := range iterations {
+		func() {
+			defer func() {
+				if r := recover(); r != nil {
+					t.Errorf("iteration %d: panic in LoginCommand.Run (would crash tofu login): %v", i, r)
+				}
+			}()
+
+			workDir := t.TempDir()
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			loginView, done := testView(t)
+			defer done(t)
+
+			creds := cliconfig.EmptyCredentialsSourceForTests(
+				filepath.Join(workDir, "credentials.tfrc.json"),
+			)
+			svcs := disco.New(
+				disco.WithCredentials(creds),
+				disco.WithHTTPClient(httpclient.New(ctx)),
+			)
+			svcs.ForceHostServices(svchost.Hostname("example.com"), map[string]any{
+				"login.v1": map[string]any{
+					"client": "anything-goes",
+					"authz":  s.URL + "/authz",
+					"token":  s.URL + "/token",
+				},
+			})
+
+			abortCh := make(chan struct{})
+			c := &LoginCommand{
+				Meta: Meta{
+					WorkingDir:      workdir.NewDir("."),
+					View:            loginView,
+					BrowserLauncher: webbrowser.NewMockLauncher(ctx),
+					Services:        svcs,
+					ShutdownCh:      abortCh,
+				},
+			}
+
+			defer testInputMap(t, map[string]string{
+				"approve": "yes",
+			})()
+
+			statusCh := make(chan int, 1)
+			go func() {
+				statusCh <- c.Run([]string{"example.com"})
+			}()
+
+			// Fire ShutdownCh at varying delays to hit different timing windows:
+			// - delay 0: ShutdownCh fires before MockLauncher visits the callback URL
+			// - small delay: races with the HTTP callback handler
+			// - larger delay: fires after OAuth is already complete
+			go func() {
+				time.Sleep(time.Duration(i%5) * time.Millisecond)
+				close(abortCh)
+			}()
+
+			select {
+			case <-statusCh:
+				// exit status 0 (OAuth won) or 1 (ShutdownCh won) are both valid
+			case <-time.After(10 * time.Second):
+				t.Errorf("iteration %d: LoginCommand.Run did not return — possible deadlock", i)
+			}
+		}()
+	}
+}
+
+// TestLoginOAuthCallbackNoPanicOnAbort verifies that sending to ShutdownCh
+// before the OAuth callback arrives causes a clean abort without deadlock.
+func TestLoginOAuthCallbackNoPanicOnAbort(t *testing.T) {
+	s := httptest.NewServer(oauthserver.Handler)
+	defer s.Close()
+
+	for i := range 50 {
+		func() {
+			defer func() {
+				if r := recover(); r != nil {
+					t.Errorf("iteration %d: unexpected panic: %v", i, r)
+				}
+			}()
+
+			workDir := t.TempDir()
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			loginView, done := testView(t)
+			defer done(t)
+
+			creds := cliconfig.EmptyCredentialsSourceForTests(
+				filepath.Join(workDir, "credentials.tfrc.json"),
+			)
+			svcs := disco.New(
+				disco.WithCredentials(creds),
+				disco.WithHTTPClient(httpclient.New(ctx)),
+			)
+			svcs.ForceHostServices(svchost.Hostname("example.com"), map[string]any{
+				"login.v1": map[string]any{
+					"client": "anything-goes",
+					"authz":  s.URL + "/authz",
+					"token":  s.URL + "/token",
+				},
+			})
+
+			// No MockLauncher: the OAuth callback will never arrive, so ShutdownCh
+			// is the only way to unblock the command.
+			abortCh := make(chan struct{})
+			c := &LoginCommand{
+				Meta: Meta{
+					WorkingDir: workdir.NewDir("."),
+					View:       loginView,
+					Services:   svcs,
+					ShutdownCh: abortCh,
+				},
+			}
+
+			defer testInputMap(t, map[string]string{
+				"approve": "yes",
+			})()
+
+			statusCh := make(chan int, 1)
+			go func() {
+				statusCh <- c.Run([]string{"example.com"})
+			}()
+
+			close(abortCh)
+
+			select {
+			case status := <-statusCh:
+				if status != 1 {
+					t.Errorf("iteration %d: expected exit status 1 after abort, got %d", i, status)
+				}
+			case <-time.After(10 * time.Second):
+				t.Errorf("iteration %d: LoginCommand.Run did not return after ShutdownCh — deadlock", i)
+			}
+		}()
+	}
 }

